### PR TITLE
Luis/nlp classifier module

### PR DESCRIPTION
**Pull request by LDiazN**
# Problem 
We need a starting architecture for our classifier, so we can easily do some experiments. Such classifier is implemented in a POC by @dieko95 that you can find [here](https://github.com/code-for-venezuela/c4v-py/pull/77), here we are just integrating that implementation to our current project

# Solution
* Write a class that automates the entire pipeline for this classifier, the main target is that such class implements this process in multiple steps that can be easily changed or configured using parameters
*  Use a scheme for managing multiple experiments, so we can manage multiple results: experiments are defined by a branch name and an experiment name, in a folder under the `.c4v` folder at `$HOME` by default

# Relevant files
* `src/c4v/classifier/classifier.py` : Here we defined the `ClassifierExperiment` class that automates the entire process of a classification training. The main function is `run` and you can pass it a dict with the fields for training arguments to override the default settings

# How to test it 
* Create a script, say `test.py` at the root project folder 
* Write the folowwing code:
```
from c4v.classifier.classifier import ClassifierExperiment

# branch name, experiment name
experiment = ClassifierExperiment("testing", "first_one")

print(experiment.run(train_args={'num_train_epochs' : 3}))
```
* Run script with python
* The following results were obtained from google colab for the previous experiment:
```
Training completed. Do not forget to share your model on huggingface.co/models =)


{'train_runtime': 1051.2505, 'train_samples_per_second': 5.479, 'train_steps_per_second': 0.548, 'train_loss': 0.6006555491023593, 'epoch': 3.0}
100% 576/576 [17:31<00:00,  1.83s/it]
Configuration saved in /experiments/testing/first_one/config.json
Model weights saved in /experiments/testing/first_one/pytorch_model.bin
***** Running Evaluation *****
  Num examples = 480
  Batch size = 10
100% 48/48 [00:30<00:00,  1.58it/s]
                         metrics_value
eval_loss                     0.429582
eval_accuracy                 0.797917
eval_precision                0.775701
eval_recall                   0.772093
eval_f1                       0.773893
eval_runtime                 31.039800
eval_samples_per_second      15.464000
eval_steps_per_second         1.546000
epoch                         3.000000
```

# Further work
* Do some more experiments to improve classifications
* Add a class to instantiate a classifier from an experiment 
* Add configuration manager to handle configuration variables as the `BASE_C4V_FOLDER` variable in `classifier.py`

### CODE REVIEW

File: src/c4v/classifier/classifier.py [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+import torch
+import pandas as pd
+import numpy as np
+import os
+
+from transformers.trainer_utils import EvalPrediction
+
+device = torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")
+
+BASE_C4V_FOLDER = os.path.join(os.environ.get("HOME"), "/.c4v")
```
Comment (Edilmo) [inline src/c4v/classifier/classifier.py:30]: What about having all paths as settings?

File: src/c4v/classifier/classifier.py [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+BASE_C4V_EXPERIMENTS_FOLDER = os.path.join(BASE_C4V_FOLDER, "/experiments")
+
+class Tags(Enum):
+    """
+        Possible tags variants for an article
+    """
+    DENUNCIA_FALTA_DEL_SERVICIO = "DENUNCIA FALTA DEL SERVICIO"
+
+class ClassifierExperiment:
```
Comment (Edilmo) [inline src/c4v/classifier/classifier.py:39]: I think this class could be a Generic.

We can keep it as is for now to move forward but let me shed some light:

1. If we want to classify with more than one label (multi class instead of binary) the code is almost all the same.
2. If we want to use a different HF model, we just need to change the model,and the tokenizer (two methods)
3. This is for Text classification which could be reflected in the class name to keep the door open for image classification for example.
4. The text classification experiment could be a different class that consumes a text clasifier, which allows the experiment to define the dataset.

File: test_file.py [hunk 3]
```diff
@@ -0,0 +1,6 @@
+from c4v.classifier.classifier import ClassifierExperiment
+
+# branch name, experiment name
```
Comment (Edilmo) [inline test_file.py:3]: Are we sure we want to put this file in the root folder?

**Review by Edilmo — APPROVED** [review]
I'm approving to let you know when and how address my suggestions.

File: src/c4v/classifier/classifier.py (hunk 1 above)
Comment (LDiazN) [reply src/c4v/classifier/classifier.py:30]: Yes, I have to change it, this was done before having dynaconf settings :)

File: test_file.py (hunk 3 above)
Comment (LDiazN) [reply test_file.py:3]: oh, I have to remove it, I push it just as a workaround in google colab, as I wasn't able to run it because of the python version
